source/common/tracing/http_tracer_manager_impl.cc

```diff
@@ -0,0 +1,39 @@
+#include "common/tracing/http_tracer_manager_impl.h"
+
+#include "common/config/utility.h"
+
+namespace Envoy {
+namespace Tracing {
+
+HttpTracerManagerImpl::HttpTracerManagerImpl(
+    Server::Configuration::TracerFactoryContextPtr factory_context)
+    : factory_context_(std::move(factory_context)) {}
+
+HttpTracerSharedPtr
+HttpTracerManagerImpl::getOrCreateHttpTracer(const envoy::config::trace::v3::Tracing_Http* config) {
+  if (!config) {
+    return null_tracer_;
+  }
+
+  const auto cache_key = MessageUtil::hash(*config);
+  const auto it = http_tracers_.find(cache_key);
+  if (it != http_tracers_.end()) {
+    return it->second;
+  }
+
+  // Initialize tracing driver.
+  ENVOY_LOG(info, "instantiating tracing driver: {}", config->name());
+
+  // Now see if there is a factory that will accept the config.
+  auto& factory =
+      Envoy::Config::Utility::getAndCheckFactory<Server::Configuration::TracerFactory>(*config);
+  ProtobufTypes::MessagePtr message = Envoy::Config::Utility::translateToFactoryConfig(
+      *config, factory_context_->messageValidationVisitor(), factory);
+
+  HttpTracerSharedPtr http_tracer = factory.createHttpTracer(*message, *factory_context_);
+  http_tracers_.emplace(cache_key, http_tracer);
```

Comment:
How does anything ever got removed from this map? It seems like a tracer that is no longer references should be cleaned up?

Reply:
I'd like to leave clean-up logic out of scope for this PR and focus on refactoring of the existing bootstrap tracing config.

But it's probably a good time to discuss the next steps necessary.

A couple observations:
1. None of the existing tracers are taking any cleanup actions right now (well, probably `lightstep` is trying)
2. I don't know if `dynamic_ot` and `opencensus` can even support cleanup in general (since actual tracer implementations are not part of Envoy code base)

So, I would suggest first to add support for cleanup into tracers, and then revise `HttpTracerManager`.

Is it ok ?

Reply:
Apart from cleanup I'm mainly just talking about freeing memory. In the contrived case in which the user keeps changing the tracing config, the old tracers will never get freed. The general pattern here would be to store a weak_ptr in the map and do some cleanup when the reference is actually gone. I think it's fine to do this in a follow up but can you add a TODO/comment here about this?

Reply:
Sure, I will add a TODO

Reply:
added